src/cli.ts

```diff
@@ -0,0 +1,141 @@
+ */
+
+import * as os from 'os';
+import * as path from 'path';
+import { Config, Interfaces, run as oclifRun } from '@oclif/core';
+import { set } from '@salesforce/kit';
+import { AnyJson, get } from '@salesforce/ts-types';
+import * as Debug from 'debug';
+import { default as nodeEnv, Env } from './util/env';
+
+const debug = Debug('sf');
+
+export const UPDATE_DISABLED_INSTALLER =
+  'Manual and automatic CLI updates have been disabled by setting "SF_AUTOUPDATE_DISABLE=true". ' +
+  'To check for a new version, unset that environment variable.';
+
+export const UPDATE_DISABLED_NPM = 'Use "npm update --global sfdx-cli" to update npm-based installations.';
+
+export const UPDATE_DISABLED_DEMO =
+  'Manual and automatic CLI updates have been disabled in DEMO mode. ' +
+  'To check for a new version, unset the environment variable SF_ENV.';
+
+export function configureUpdateSites(config: Interfaces.Config, env = nodeEnv): void {
+  const s3Host = env.getS3HostOverride();
+  if (s3Host) {
+    // Override config value if set via envar
+    set(config, 'pjson.oclif.update.s3.host', s3Host);
+  }
+
+  const npmRegistry = env.getNpmRegistryOverride();
+  if (npmRegistry) {
+    // Override config value if set via envar
+    set(config, 'pjson.oclif.warn-if-update-available.registry', npmRegistry);
+  }
+}
+
+export function configureAutoUpdate(envars: Env): void {
+  if (envars.isDemoMode()) {
+    // Disable autoupdates in demo mode
+    envars.setAutoupdateDisabled(true);
+    envars.setUpdateInstructions(UPDATE_DISABLED_DEMO);
+    return;
+  }
+
+  if (envars.isInstaller()) {
+    envars.normalizeAutoupdateDisabled();
+    if (envars.isAutoupdateDisabled()) {
+      envars.setUpdateInstructions(UPDATE_DISABLED_INSTALLER);
+    }
+    return;
+  }
+
+  // Not an installer, so this must be running from an npm installation
+  if (!envars.isAutoupdateDisabledSet()) {
+    // Disable autoupdates if run from an npm install or in local dev, if not explicitly set
+    envars.setAutoupdateDisabled(true);
+  }
+
+  if (envars.isAutoupdateDisabled()) {
+    envars.setUpdateInstructions(UPDATE_DISABLED_NPM);
+  }
+}
+
+function debugCliInfo(version: string, channel: string, env: Env, config: Interfaces.Config): void {
+  function debugSection(section: string, items: Array<[string, string]>): void {
+    const pad = 25;
+    debug('%s:', section.padStart(pad));
+    items.forEach(([name, value]) => debug('%s: %s', name.padStart(pad), value));
+  }
+
+  debugSection('OS', [
+    ['platform', os.platform()],
+    ['architecture', os.arch()],
+    ['release', os.release()],
+    ['shell', config.shell],
+  ]);
+
+  debugSection('NODE', [['version', process.versions.node]]);
+
+  debugSection('CLI', [
+    ['version', version],
+    ['channel', channel],
+    ['bin', config.bin],
+    ['data', config.dataDir],
+    ['cache', config.cacheDir],
+    ['config', config.configDir],
+  ]);
+
+  debugSection(
```

Comment:
Not a requirement but it might be nice to list out every environment variable that starts with `SF` or `SFDX`

e.g.

```typescript
const vars = Object.keys(process.env).filter(e => e.startsWith('SF_') || e.startsWith('SFDX_'))
```